### PR TITLE
Added logging for CallerProcedure header

### PR DESCRIPTION
**Pull request by gopalchouhan**
CallerProcedure header was added as part of [pull request-2027](https://github.com/yarpc/yarpc-go/pull/2027).
This pull request is the follow up request and takes care of specifically logging for the CallerProcedure header.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,7 +5,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
-- No changes yet.
+- Added logging for CallerProcedure header
```
Comment (Dogild) [inline CHANGELOG.md:8]: nit: this should be

```
## [Unreleased]
### Added
- observability: added key `sourceProcedure` in logging. The value of the key will be the `CallerProcedure` value of the request.
```

File: internal/observability/graph.go [hunk 2]
```diff
@@ -105,6 +105,7 @@ func (g *graph) begin(ctx context.Context, rpcType transport.Type, direction dir
 	d.Add(req.Procedure)
 	d.Add(req.RoutingKey)
 	d.Add(req.RoutingDelegate)
+	d.Add(req.CallerProcedure)
```
Comment (jronak) [inline internal/observability/graph.go:108]: I'm concerned about creating a new edge for every different callProcedure and other keys combinations (routing delegate, routing key) as this will increase edge cache size significantly for services with different routing-key/routing-delegate and also it might put pressure on the m3 collector as it's going to receive too unaggregated metrics. 

An alternative option might be to pass source-procedure in the zap extra fields accepted by the `endWithAppError` method, what do you think?

https://github.com/yarpc/yarpc-go/blob/9d0cab62f323aac0e60616fd3952d5b36ca85479/internal/observability/call.go#L264

Comment (gopalchouhan) [reply internal/observability/graph.go]: Very good point! given that this change is just for logging, we don't want to include CallerProcedure in to the key generation.

**Review by Dogild — APPROVED** [review]

File: internal/observability/call.go [hunk 3]
```diff
@@ -223,6 +231,7 @@ func (c call) endLogs(
 	}
 
 	fields := c.fields[:0]
+	fields = c.addMoreFieldsFromTransportRequest(fields)
```
Comment (Dogild) [inline internal/observability/call.go:234]: I am not sure to understand exactly why you added a new methods `c.addMoreFieldsFromTransportRequest` for this? Why not simply having:

```
fields := c.fields[:0]
fields = append(fields, zap.String("sourceProcedure", c.req.CallerProcedure))
fields = append(fields, zap.String("rpcType", c.rpcType.String()))
```

I can see you are also using the same method in the streaming log functions, but having a function here seems a bit extreme and is also a different pattern than the current way we add those fields (like rpcType). Did I miss something?

File: internal/observability/call.go [hunk 4]
```diff
@@ -483,12 +492,13 @@ func (c call) logStreamEvent(err error, success bool, succMsg, errMsg string, ex
 		ce = c.edge.logger.Check(lvl, errMsg)
 	}
 
-	fields := []zap.Field{
-		zap.String("rpcType", c.rpcType.String()),
+	fields := c.fields[:0]
```
Comment (Dogild) [inline internal/observability/call.go:495]: Can we do this in another PR? This seems to add more logs for streaming and is out of scope of this diff. (but I appreciated this fix though!)

File: internal/observability/middleware_test.go [hunk 5]
```diff
@@ -589,6 +594,89 @@ func TestMiddlewareLoggingWithApplicationErrorConfiguration(t *testing.T) {
...
+		return e
+	}
+
+	wantFields := []zapcore.Field{
+		zap.Duration("latency", 0),
+		zap.Bool("successful", true),
+		zap.Skip(), // ContextExtractor
+	}
+
+	for i := 0; i < 2; i++ {
```
Comment (Dogild) [inline internal/observability/middleware_test.go:648]: please use same testing pattern as other tests for this:

```
	type test struct {
		callerProcedure string
		wantFields []zapcore.Field
	}
        tests := []test{
          {
             callerProcedure: "caller-procedure-1",
             wantFields: []zapcore.Field{...}
          },
                    {
             callerProcedure: "caller-procedure-2"
             wantFields: []zapcore.Field{...}
          },
        }
        ...
```

File: internal/observability/middleware_test.go [hunk 6]
```diff
@@ -589,6 +594,89 @@ func TestMiddlewareLoggingWithApplicationErrorConfiguration(t *testing.T) {
...
+				&transporttest.FakeResponseWriter{},
+				fakeHandler{},
+			)
+			assert.NoError(t, err, "Unexpected error from middleware.")
+			logContext := append(
+				baseFields(),
+				zap.String("direction", string(_directionInbound)),
+				zap.String("sourceProcedure", req.CallerProcedure),
+				zap.String("rpcType", "Unary"),
+			)
```
Comment (Dogild) [inline internal/observability/middleware_test.go:664]: Should the fields `baseFields()`, `direction` and  `rpcType` be in `wantFields`?

File: internal/observability/call.go (hunk 4 above)
Comment (gopalchouhan) [reply internal/observability/call.go:495]: shouldn't we be adding callerProcedure logging by default for both unary and stream?

File: internal/observability/call.go (hunk 3 above)
Comment (gopalchouhan) [reply internal/observability/call.go:234]: I wanted to keep the field extraction from transport request in a single place to avoid duplicating the same code in multiple places. I did not include "rpcType" and other fields because they don't belong to transportRequest. Do you think any performance impact this may have or just the regrouping comment?
Also, do you feel very strong about this comment?

Comment (Dogild) [reply internal/observability/call.go:234]: No there will be not major performance impact from this. Heap will be the same.

It is purely esthetic for the readability of the code, with your solution people reading this will need to jump to another function (which is 1 line of code) to understand the flow. IMO readability wins in comparison with the gain of the refactoring (1 line of code).

File: internal/observability/call.go (hunk 4 above)
Comment (Dogild) [reply internal/observability/call.go:495]: Nevermind this comment, I read something else than `c.fields[:0]`

**Review by Dogild — APPROVED** [review]

File: internal/observability/call.go (hunk 3 above)
Comment (AllenLuUber) [reply internal/observability/call.go:234]: I agree with @Dogild . Also to be consistent to all the methods below, we should just do it inline like @Dogild suggested.

File: internal/observability/graph.go [hunk 7]
```diff
@@ -477,6 +478,7 @@ func newEdge(logger *zap.Logger, meter *metrics.Scope, metricTagsBlocklist []str
 		zap.String("encoding", string(req.Encoding)),
 		zap.String("routingKey", req.RoutingKey),
 		zap.String("routingDelegate", req.RoutingDelegate),
+		zap.String("sourceProcedure", req.CallerProcedure),
```
Comment (jkanywhere) [inline internal/observability/graph.go]: Why doesn't the log key match the property name?

When the `transport.Request` object is logged, the `CallerProcedure` property is logged as `callerProcedure`.
https://github.com/yarpc/yarpc-go/blob/dev/api/transport/request.go#L107

Comment (jkanywhere) [reply internal/observability/graph.go]: Also the on-the-wire representation is `rpc-caller-procedure` , not `source-procedure`, so I think `callerProcedure` makes more sense in the logs.

https://github.com/yarpc/yarpc-go/blob/9d0cab62f323aac0e60616fd3952d5b36ca85479/transport/grpc/headers.go#L40
https://github.com/yarpc/yarpc-go/blob/f83a481b968915f1d834d2e66daf799ccd521361/transport/tchannel/header.go
https://github.com/yarpc/yarpc-go/blob/f83a481b968915f1d834d2e66daf799ccd521361/transport/http/constants.go#L45

**Comment by Dogild** [comment]
Sync with @gopalchouhan and this will not be landed in yarpc
